tests/e2e/e2e_setup_test.go

```diff
@@ -391,84 +391,116 @@ func noRestart(config *docker.HostConfig) {
 	}
 }
 
-func (s *IntegrationTestSuite) initUpgrade() {
+func (s *IntegrationTestSuite) upgrade() {
 	// submit, deposit, and vote for upgrade proposal
-	s.submitProposal(s.chains[0])
-	s.submitProposal(s.chains[1])
+	currentHeightA := s.chainHeight(s.valResources[s.chains[0].ChainMeta.Id][0].Container.ID)
+	currentHeightB := s.chainHeight(s.valResources[s.chains[1].ChainMeta.Id][0].Container.ID)
+	propHeightA = currentHeightA + 40 + len(s.chains[0].Validators)
```

Comment:
Is it possible to instead programmatically force the chain to pause at a certain height so that we can deterministically supply the desired upgrade height?

I don't know how to do it or if it is possible. The reason for this suggestion is to prevent having non-determinism such as where `currentHeightA` is different from each test run. That is likely to cause bugs in the future. From my experience, such things are very painful to debug. 

Also, the choice of 40. It might work today but not tomorrow if block time changes. That's why programmatically enforcing a certain height might be more stable long-term

Reply:
Yes, after thinking about this for a while I think the method you suggested is better. What I liked about this method is it forced the chain to prove that it halted at the specified height, but what we can instead do is halt at the upgrade height and if on upgrade we dont hit blocks we can assume something wrong happened with the proposal

Reply:
Via DMs with @p0mvn 

Decided to more deterministically predict how long it takes to create, deposit, and vote for a proposal. This will translate into a variable voting period that will get created via a flag in the init container